Exploration for a data layer

altendky opened the pull request:
Draft for:
- [ ] Being more than an exploration
- [ ] Moving to a new data layer repository

altendky committed:
Preliminary exploration of a data layer skeleton

altendky commented:
Alrighty.  Should I look to https://github.com/chia-network/chiapos for layout and CI reference?

A quick grep didn't turn up any imports of `chia` in `chiapos`, but I already am referencing some tooling located in chia-blockchain here such as `chia.types.blockchain_format.sized_bytes.bytes32` and `chia.util.db_wrapper.DBWrapper`.  Are we just ok with bidirectional dependencies between repos, or is this a sign I'm doing something wrong?  I suppose really that it's unlikely to be exclusively 'here' or 'there' but rather both in which case maybe it's ok to develop here until we settle on what code belongs where and deal with the extra CI work then?

Or, I may just be thinking about things entirely aside from your reasons.  I'm all for granular repositories.  That often leads to breaking other things down as well like putting utilities in a 'third' project which has corresponding overhead either in more pre-design to avoid frequent updates or just dealing with frequent updates and compatibility.  I have no history here to know how those decisions have been or will be made.

hoffmang9 commented:
I think the implied process being proposed here is to make chia-blockchain a sub repo of a datalayer repo.

altendky commented:
Hmm, ok.  I was thinking of data layer as another service to run but that sounds more like a separate application that happens to use some existing bits in `chia-blockchain`.  I'll adjust my perspective.

Perhaps it is still ok to leave the extraction and CI setup for a bit later.  I noted it in the OP.

altendky committed:
add db table creation and a trivial test

altendky committed:
black

altendky committed:
black and flake8

altendky committed:
more for pylint

altendky committed:
fill out more of data_layer.py for the linter

mariano54 commented:
The data layer is a service, kind of like the pool is a service, and layer two lighning nodes will be services. However, they are not part of the core blockchain (layer 1), and therefore should be a new repo.
I like the structure of using chia-blockchain by including a subrepo like Gene said, and then requiring the node to be running in order to be able to run the service.
One key difference with these layer 2 services (vs services in chia-blockchain) is that they don't use the Chia protocol (websockets, handshake, node ids, etc), and instead just use the RPC, since they are most separated. This keeps the code simpler.

altendky commented:
Thanks for the explanation.  I'll take a look at https://github.com/chia-network/pool-reference for familiarity.

hoffmang9 commented:
@cmmarslender pointed out a potentially much better way to depend on chia-blockchain as he was cleaning up chiadns. Instead of submoduling chia-blockchain, just make it a dependency in setup.py for the datalayer.

mariano54 commented:
> @cmmarslender pointed out a potentially much better way to depend on chia-blockchain as he was cleaning up chiadns. Instead of submoduling chia-blockchain, just make it a dependency in setup.py for the datalayer.

oh you're right! That is better. That's what I meant, it's how pool reference works.

almogdepaz commented on chia/data_layer/data_store.py:
```diff
@@ -0,0 +1,138 @@
+
+        return self
+
+    # TODO: Add some handling for multiple tables.  Could be another layer of class
```
we can make each row can contain a "Table" column containing the id of the table it belongs to.
that way we can manage adding an removing tables easy

altendky replied:
Agreed, I meant multiple data layer tables, not multiple sql tables.

almogdepaz committed:
start datalayer rpc and wallet

almogdepaz committed:
tidy imports

altendky committed:
Hint as CLVMObject.CLVMObject for now

altendky committed:
fix up CLVMObject hints

altendky committed:
implement some data layer DB insertion and deletion

altendky committed:
black

altendky committed:
flake8

altendky committed:
black

altendky commented on chia/rpc/rpc_server.py:
```diff
@@ -101,14 +101,14 @@ async def get_connections(self, request: Dict) -> Dict:
         if "node_type" in request:
             request_node_type = NodeType(request["node_type"])
-        if self.rpc_api.service.server is None:
+        if self.rpc_api.data_layer_service.server is None:
```
@almogdepaz, it looks like maybe a refactor rename reached a bit far out?  Though I think maybe this code expects the RPC implementations to use `.service`?

altendky replied:
I suppose I ought to link to an example error in the CI runs.

https://github.com/Chia-Network/chia-blockchain/pull/8492/checks?check_run_id=3646582478#step:14:729
```
rpc_api = <chia.rpc.full_node_rpc_api.FullNodeRpcApi object at 0x7f0400847370>
self_hostname = 'localhost', daemon_port = 55400, rpc_port = 51000
stop_cb = <bound method Service.stop of <chia.server.start_service.Service object at 0x7f0401b15220>>
root_path = PosixPath('/tmp/tmpt5azhi35')
net_config = {'ALERTS_URL': 'download.chia.net/notify/mainnet_alert.txt', 'CHIA_ALERTS_PUBKEY': '89b7fd87cb56e926ecefb879a2...a': {'crt': 'config/ssl/ca/chia_ca.crt', 'key': 'config/ssl/ca/chia_ca.key'}, 'daemon_max_message_size': 50000000, ...}
connect_to_daemon = False

    async def start_rpc_server(
        rpc_api: Any,
        self_hostname: str,
        daemon_port: uint16,
        rpc_port: uint16,
        stop_cb: Callable,
        root_path: Path,
        net_config,
        connect_to_daemon=True,
    ):
        """
        Starts an HTTP server with the following RPC methods, to be used by local clients to
        query the node.
        """
        app = aiohttp.web.Application()
        rpc_server = RpcServer(rpc_api, rpc_api.service_name, stop_cb, root_path, net_config)
>       rpc_server.rpc_api.data_layer_service._set_state_changed_callback(rpc_server.state_changed)
E       AttributeError: 'FullNodeRpcApi' object has no attribute 'data_layer_service'
```

almogdepaz committed:
fix naming

almogdepaz commented on chia/data_layer/data_store.py:
```diff
@@ -108,29 +109,68 @@ async def get_row_by_index(self, index: int) -> CLVMObject:
         pass
 
-    async def insert_row(self, table_id: bytes32, clvm_object: CLVMObject) -> None:
+    async def insert_row(self, table: bytes32, clvm_object: CLVMObject, index: Optional[int] = None) -> None:
```
does the user determine the row index  on submission ? 
i think that it should just be a running id

altendky replied:
All the examples had `Row #1` etc so I added indexes.  As I look back, I may well have been reading too much into that and it may be removed.  It is a point to discuss as to whether the rows have order to them.

altendky committed:
remove some manual/auto-formatting silliness

altendky committed:
add DataStore.get_rows_by_hash()

altendky committed:
Merge branch 'main' into altendky-data_layer_exploration

altendky committed:
undo formatting change

altendky committed:
remove some commented out test code

altendky committed:
make chia.data_layer a properly installed package

altendky committed:
stricter mypy, add DataStore.get_row_by_index(), introduce TableRow

Compound commits are bad.  Making mypy stricter pointed out some things and
they snowballed.  If this is problematic even at this early stage, I can
adjust and put more effort into avoiding this.

altendky committed:
add missing db.commit()

altendky committed:
add data layer action logging

altendky committed:
Merge branch 'main' into altendky-data_layer_exploration

altendky committed:
lint

altendky committed:
formatting

altendky committed:
Merge branch 'main' into altendky-data_layer_exploration

altendky committed:
define foreign keys for data layer

altendky committed:
entirely remove index from data layer code

altendky committed:
use sets instead of lists in data layer tests, order was removed

altendky committed:
switch data layer to simple key value store, no order/indexes or dpulicates

altendky committed:
only import, not also from

altendky committed:
@dataclass for DataStore

altendky committed:
use aiosqlite.Row to access elements by name in data layer

altendky committed:
more async for in data layer

altendky committed:
explicit .hex()/bytes32(bytes.fromhex()) for data store hashes

altendky committed:
preliminary data layer cli

almogdepaz commented on chia/cmds/chia.py:
```diff
@@ -118,6 +119,7 @@ def run_daemon_cmd(ctx: click.Context, wait_for_unlock: bool) -> None:
 cli.add_command(netspace_cmd)
 cli.add_command(farm_cmd)
+cli.add_command(data_cmd)
 
```
ok but seems that when we split to a different repo this should change and the main chia node should not include commands related to data-layer

almogdepaz commented on chia/rpc/rpc_server.py:
```diff
@@ -302,40 +302,54 @@ async def start_rpc_server(
     Starts an HTTP server with the following RPC methods, to be used by local clients to
     query the node.
     """
```
this refactor seems unrelated to the data-layer, i suggest you cherry pick it into another pr

altendky replied:
(discussed elsewhere, for the record here...)  It certainly is, it was added to identify a problem I ran into here so I kept it around.  But yes, if committed to chia-blockchain it would be via a different PR for just this.

altendky committed:
outdated comment cleanup

almogdepaz commented on chia/data_layer/data_store.py:
```diff
@@ -0,0 +1,334 @@
+        # TODO: make this handle multiple data layer tables
+        # TODO: do we need to handle multiple equal rows
+
+        await self.db.execute("CREATE TABLE IF NOT EXISTS tables(id TEXT PRIMARY KEY, name STRING)")
```
they does a table need two identifiers (name and id ?)

altendky replied:
No, that's just me going beyond where I should and presuming that a human readable name would be good someplace.  Though, it probably belongs on the NFT or such if anywhere.

altendky committed:
add a basic data layer cli integration test

altendky committed:
mypy

altendky committed:
correct data cli test to use subprocess.TimeoutExpired

altendky committed:
add DBWrapper.locked_transaction(), use in data store

altendky committed:
flake8

almogdepaz committed:
dedup code in data funcs

almogdepaz committed:
naming

almogdepaz committed:
refactor update table

altendky committed:
Correct `DataStore.add_action()` recursive mis-call

Co-authored-by: Almog De Paz <almogdepaz@gmail.com>

altendky committed:
Merge pull request #8653 from Chia-Network/altendky-data_layer_exploration-locked_transactions

altendky committed:
todo cleanup

altendky committed:
add data store table id foreign key to actions

altendky committed:
constrain all data store columns NOT NULL

altendky committed:
add generated data layer ci workflows

altendky committed:
PathLike and CompletedProcess only generic for type checking

altendky committed:
Merge pull request #8692 from Chia-Network/altendky-data_layer_exploration-add_tests_to_ci

altendky committed:
start daemon/data for data integration test

altendky committed:
non-zero exit code for data cli commands if query fails

altendky committed:
data layer typo fix

altendky committed:
return response or None from all data cli functions

altendky committed:
fixup data layer cli test

altendky committed:
add log reporting for data layer cli test

altendky committed:
mypy

altendky committed:
Merge branch 'altendky-data_layer_exploration' into altendky-data_layer_exploration-just_serialized_program

altendky committed:
use SerializedProgram in the DataStore interface

altendky committed:
Merge branch 'main' into altendky-data_layer_exploration

altendky committed:
misc data layer tidying

altendky committed:
mypy

altendky committed:
Merge branch 'altendky-data_layer_exploration' into altendky-data_layer_exploration-just_serialized_program

altendky committed:
Merge pull request #8699 from Chia-Network/altendky-data_layer_exploration-just_serialized_program

altendky committed:
Merge branch 'main' into altendky-data_layer_exploration

altendky committed:
rewrite data layer based around a CLVM object tree (#8738)

* end of day for a start into the set -> tree data layer transition

* data layer node first/rest split to left/right and key/value

* data layer can maybe insert into a tree

* Apply suggestions from code review

* add DataStore.get_tree_as_program(), and misc

* pass data store balanced 4KV build

* data store query parent -> tree from root hash

* DataStore.get_heritage() (and misc)

* DataStore.get_pairs()

* data store fail on insert of already present key

almogdepaz committed:
lint

almogdepaz committed:
unused imports

almogdepaz committed:
data_layer_types.py
data_layer_util.py
add tree id to nodes table

almogdepaz committed:
revert get_raw_pairs

almogdepaz committed:
lint

altendky committed:
catch up data store with some refactors

altendky committed:
lint

altendky committed:
remove no longer used data store test data

altendky committed:
'simplify' data layer test example data definition

altendky committed:
add lock parameter to DbWrapper.locked_transaction()

altendky committed:
make DataStore.get_node_by_key[_bytes]() tree id aware

altendky committed:
lint

altendky committed:
preliminary DataStore.delete()

altendky committed:
Merge branch 'main' into altendky-data_layer_exploration

almogdepaz committed:
Update tests/core/data_layer/test_data_store.py

remove comment

Co-authored-by: Kyle Altendorf <sda@fstab.net>

almogdepaz committed:
fix bugs in rpc_api

almogdepaz committed:
Merge branch 'altendky-data_layer_exploration' of github.com:Chia-Network/chia-blockchain into altendky-data_layer_exploration

almogdepaz committed:
add wallet to data_layer.py

altendky committed:
drop hexstr_to_bytes32()

altendky committed:
remove outdated comment

altendky committed:
lint

altendky committed:
expand DataStore.delete() testing

almogdepaz committed:
cli, some refactoring

altendky committed:
correct test_data_store.test_delete_from_left_both_terminal

altendky committed:
rework DataStore.get_heritage() to .get_ancestors()

altendky committed:
remove a stray `r`

altendky committed:
factor out to DataStore._insert_root()

altendky committed:
xfail test_data_store.test_create_tree_fails_for_not_bytes32 for now

altendky committed:
factor out to DataStore._insert_internal_node() and ._insert_terminal_node()

altendky committed:
group DataStore._insert_*()

altendky committed:
add TODO:

altendky committed:
cleanup DataStore.insert() and .delete()

altendky committed:
Add some initial DataStore integrity checks

altendky committed:
lint

altendky committed:
" not """

altendky committed:
add DataStore._check_roots_are_incrementing()

altendky committed:
remove DataStore tree table

altendky committed:
data store tidy

altendky committed:
data store tidy

altendky committed:
centralize hashing into DataStore._insert_internal_node()

altendky committed:
centralize hashing into DataStore._insert_terminal_node()

altendky committed:
add DataStore._check_hashes()

altendky committed:
tidy

altendky committed:
Merge branch 'main' into altendky-data_layer_exploration

altendky committed:
isort

almogdepaz committed:
add delete to rpc test

almogdepaz committed:
lint

altendky committed:
shift data layer fixtures to conftest.py for reuse

almogdepaz committed:
tests

almogdepaz committed:
rpc test

almogdepaz committed:
lint

altendky committed:
move data layer add_*_example() to test utils

altendky committed:
add data layer create_example fixture

altendky committed:
make DataStore.get_pairs() return ordered by depth then left-to-right

altendky committed:
lint

altendky committed:
INSERT OR IGNORE INTO for DataStore._insert_terminal_node()

altendky committed:
mypy

altendky committed:
add DataStore.autoinsert()

altendky committed:
data layer keys and values are opaque bytes

altendky committed:
Merge branch 'main' into altendky-data_layer_exploration

altendky committed:
pragma synchronous=FULL for the data layer

altendky committed:
data store tidy

altendky committed:
black
